common/src/main/java/io/netty/util/ResourceLeakDetector.java

```diff
@@ -367,29 +380,46 @@ public void record(Object hint) {
             record0(hint);
         }
 
+        /**
+         * This method works by exponentially backing off as more records are present in the stack. Each record has a
+         * 1 / 2^n chance of dropping the top most record and replacing it with itself. This has a number of convenient
+         * properties.  First, the first record is always recorded.  Second, the very last access will always be
+         * recorded.  Third, an arbitrary number of accesses can be accepted, rather than just the last few.  Fourth,
+         * it is easy to keep a precise record of the number of elements in the stack, since each element has to know
+         * how tall the stack is.
+         *
+         * In this particular implementation, there are also some advantages. A thread local random is used to decide
+         * if something should be recorded. This means that if there is a deterministic access pattern, it is now
+         * possible to see what other accesses occur, rather than always dropping them. Second, there is roughly a
+         * linear ramp up to {@link #TARGET_RECORDS}, after which backoff occurs.  This matches typical access patterns,
+         * where there are either a high number of accesses (i.e. a cached buffer), or low (an ephemeral buffer), but
+         * not many in between.
+         *
+         * The use of atomics avoids serializing a high number of accesses, when most of the records will be thrown
+         * away.  High contention only happens when there are very few existing records, which is only likely when the
+         * object isn't shared! If this is a problem, the loop can be aborted and the record dropped, because another
+         * thread won the race.
+         */
         private void record0(Object hint) {
-            // Check MAX_RECORDS > 0 here to avoid similar check before remove from and add to lastRecords
-            if (head != null && MAX_RECORDS > 0) {
-                synchronized (head) {
-                    if (tail == null) {
-                        // already closed
+            // Check TARGET_RECORDS > 0 here to avoid similar check before remove from and add to lastRecords
+            if (TARGET_RECORDS > 0) {
+                Record oldHead;
+                Record prevHead;
+                Record newHead;
+                boolean dropped;
+                do {
+                    if ((prevHead = oldHead = headUpdater.get(this)) == null) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Extract the assignment out of this expression. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AAssignmentInSubExpressionCheck)

Reply:
ignore this one